Fixed Crop resize support

SundarRajan28 opened the pull request.

SundarRajan28 committed:
Adding fixed crop resize support

SundarRajan28 committed:
Merge branch 'ROCm:develop' into crop_resize_support

SundarRajan28 committed:
Adding correct length for iterator

SundarRajan28 committed:
Adding diffusion dataloader example

fiona-gladwin commented on rocAL/include/api/rocal_api_augmentation.h:
```diff
@@ -156,6 +156,33 @@ extern "C" RocalTensor ROCAL_API_CALL rocalCropResizeFixed(RocalContext context,
+ * \ingroup group_rocal_augmentations
+ * \param [in] context Rocal context
+ * \param [in] input Input Rocal tensor
+ * \param [in] dest_height output height
```
Change the order of the args similar to the APi

fiona-gladwin commented on rocAL/include/api/rocal_api_augmentation.h:
```diff
@@ -156,6 +156,33 @@ extern "C" RocalTensor ROCAL_API_CALL rocalCropResizeFixed(RocalContext context,
+ * \param [in] output_datatype the data type of the output tensor
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalResizeCropFixed(RocalContext context, RocalTensor input,
```
What is the difference between rocalResizeCropFixed and rocalCropResizeFixed?

I think rocalCropResizeFixed will be the right name here since we are doing a crop and then reize

SundarRajan28 replied:
There is already an API named rocalCropResizeFixed

shobana-mcw replied:
What is the difference between both API's?

SundarRajan28 replied:
Discussed offline and changed the new API name to rocalROIResize

fiona-gladwin commented on rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -426,6 +426,51 @@ rocalCropResizeFixed(
 }
 
+RocalTensor ROCAL_API_CALL
+rocalResizeCropFixed(
```
Can we not use CropResizeFixed

SundarRajan28 replied:
rocalCropResizeFixed is an already existing API

fiona-gladwin replied:
Yes, but can't we just add the aditional params to the existing API instead of creating a new one

SundarRajan28 replied:
Changed the API name to rocalROIResize

fiona-gladwin commented on rocAL/source/augmentations/geometry_augmentations/node_crop_resize.cpp:
```diff
@@ -26,14 +26,17 @@ THE SOFTWARE.
     if (_node)
         return;
 
+    if(_is_random_crop) {
```
Please check the indentations

fiona-gladwin commented on rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -229,7 +229,7 @@ def __iter__(self):
 
     def __len__(self):
-        return self.iterator_length
+        return self.iterator_length // self.batch_size
```
Is this a fix?

SundarRajan28 replied:
Yes. This is used in diffusion training tqdm bar. Without this, it gives the length as the total number of images per shard

shobana-mcw replied:
But will this have an impact on other trainings? What should len return?

SundarRajan28 replied:
This is the correct value for the length of the iterator and this is what we should return. Currently, in classification trainings, we divide the length by batch size in training pipeline and print it. I can modify the code there

fiona-gladwin commented on rocAL_pybind/amd/rocal/fn.py:
```diff
@@ -493,6 +493,36 @@ def resize_crop(*inputs, resize_width=0, resize_height=0, crop_area_factor=None,
         Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
     return (crop_resized_image)
 
+def resize_crop_fixed(*inputs, resize_width=0, resize_height=0, crop_w=None, crop_h=None, crop_pos_x=None, crop_pos_y=None,
```
This is crop_Resize_fixed right?

SundarRajan28 replied:
The random crop version of the CropResize node is added as resize_crop so this version is named resize_crop_fixed.

fiona-gladwin commented on rocAL_pybind/amd/rocal/fn.py:
```diff
@@ -493,6 +493,36 @@ def resize_crop(*inputs, resize_width=0, resize_height=0, crop_area_factor=None,
+        @param output_layout (int, optional, default = types.NHWC)                         tensor layout for the augmentation output
+        @param output_dtype (int, optional, default = types.UINT8)                         tensor dtype for the augmentation output
+
+        @return    Resized and cropped Image
```
Cropped and resized image right?

fiona-gladwin commented on tests/python_api/pytorch_diffusion_loader.py:
```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2018 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Change the copyright

fiona-gladwin commented on tests/python_api/pytorch_diffusion_loader.py:
```diff
@@ -0,0 +1,112 @@
+    # Reference: CelebA Dataset loader - https://github.com/tqch/ddpm-torch/blob/master/ddpm_torch/datasets.py#L73
+    with pipe:
+        jpegs, labels = fn.readers.file(file_root=data_path)
+        rocal_device = 'cpu' if rocal_cpu else 'gpu'
```
Where is this used?

shobana-mcw commented on rocAL/include/augmentations/geometry_augmentations/node_crop_resize.h:
```diff
@@ -31,15 +31,24 @@ class CropResizeNode : public CropNode {
     void update_node() override;
 
    private:
     std::shared_ptr<RocalRandomCropParam> _crop_param;
```
Can we not add Fixed crop support in RocalRandomCropParam? 
What is the difference between RocalRandomCropParam and RocalCropParam?

SundarRajan28 replied:
RocalRandomCrop uses aspect ratio and area factor for cropping whereas RocalCropParam uses width and height for cropping. Since we have to support both crop params here, we create two params in the node and set them accordingly in the init functions

shobana-mcw replied:
Please add it as a comment.

SundarRajan28 replied:
I'm returning either crop_param or crop_fixed_param. Since both had CropParam as base class, it can be sent as CropParam shared pointer here

shobana-mcw commented on rocAL/include/meta_data/meta_node_crop_resize.h:
```diff
@@ -37,7 +37,7 @@ class CropResizeMetaNode : public MetaNode {
    private:
     void initialize();
-    std::shared_ptr<RocalRandomCropParam> _meta_crop_param;
+    std::shared_ptr<CropParam> _meta_crop_param;
```
Why meta data support is added only for fixed crop param? We should add support for random Crop also.

SundarRajan28 replied:
The support is added for both. Previously this was a RocalRandomCropParam shared ptr which only supported CropResizeNode with random crop. I changed it to CropParam which is the base class for both RocalRandomCropParam and RocalCropParam so that this meta node can be used for both cases

shobana-mcw replied:
Mention this in a comment.

shobana-mcw commented on rocAL/source/augmentations/geometry_augmentations/node_crop_resize.cpp:
```diff
@@ -81,15 +93,31 @@ void CropResizeNode::update_node() {
 }
 
 void CropResizeNode::init(float area, float aspect_ratio, float x_center_drift, float y_center_drift) {
```
Why interpolation type is not added?

SundarRajan28 replied:
Previously CropResize node used vxExtRppResizeCrop which only supports bilinear interpolation. So the init functions didn't have the interpolation type arg. In this PR, we only add the CropResizeFixed support so I didn't change the existing CropResize API. Let me know if it should be changed as well

shobana-mcw replied:
Please make that change too.

SundarRajan28 replied:
As discussed offline, will not be changing the existing APIs and the init functions in this PR

shobana-mcw commented on rocAL/source/augmentations/geometry_augmentations/node_crop_resize.cpp:
```diff
@@ -81,15 +93,31 @@ void CropResizeNode::update_node() {
     _crop_param->set_y_drift_factor(ParameterFactory::instance()->create_single_value_param(y_center_drift));
 }
 
 void CropResizeNode::init(FloatParam *area, FloatParam *aspect_ratio, FloatParam *x_center_drift, FloatParam *y_center_drift) {
```
Why interpolation type is not added?

SundarRajan28 replied:
Previously CropResize node used vxExtRppResizeCrop which only supports bilinear interpolation. So the init functions didn't have the interpolation type arg. In this PR, we only add the CropResizeFixed support so I didn't change the existing CropResize API. Let me know if it should be changed as well

SundarRajan28 replied:
As discussed offline, will not be changing the existing APIs and the init functions in this PR

shobana-mcw reviewed (commented):
Pls address the review comments and let me know.

SundarRajan28 committed:
Resolve review comments and formatting changes

SundarRajan28 commented:
@shobana-mcw I have resolved review comments and responded to the comments about crop param changes. Please review once

SundarRajan28 committed:
Adding ROIResize function

SundarRajan28 committed:
Merge branch 'develop' into crop_resize_support

spolifroni-amd committed:
commit of the RNNT topic (#249)

* commit of the RNNT topic

* Changed syntax of the URL

kiritigowda committed:
Merge branch 'develop' into crop_resize_support

kiritigowda committed:
CTest - Updates (#248)

* CTest - Updates

* Tests - Updates

* rocAL - Test

* FindFFMPEG - Add to cache

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.11.0 to 1.12.0 in /docs/sphinx (#252)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.11.0 to 1.12.0.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.11.0...v1.12.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

fiona-gladwin committed:
GPU pipeline errors - AMD Clang++ (#251)

* Set graph affinity group and reserved to 0

* Set graph and reserved to 0 in context affinity

---------

Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

SundarRajan28 committed:
Merge branch 'develop' into crop_resize_support

SundarRajan28 committed:
Resolving review comments

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.12.0 to 1.12.1 in /docs/sphinx (#257)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.12.0 to 1.12.1.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.12.0...v1.12.1)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kiritigowda committed:
Merge branch 'develop' into crop_resize_support

SundarRajan28 committed:
AMDClang++ - Fix warnings (#254)

* Remove unused variables

* Add missing override keyword

* Add virtual destructors

* Fix warnings wrt variable length arrays

* Remove unused variables and minor fix

* Fix warnings wrt moving temporary objects

* Remove unused private variables and assignment in constructor

* Fix warnings wrt return-type-c-linkage for Metadata API

* Fix warnings wrt return-type-c-linkage for NonSilentRegion

* Introduce pybind class for rocalListOfTensorList

* Fix warnings in C++ tests

Use vectors for dynamically allocated arrays

---------

Co-authored-by: fiona-gladwin <fionagladwin@multicorewareinc.com>
Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Merge branch 'develop' into crop_resize_support

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.12.1 to 1.13.0 in /docs/sphinx (#260)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.12.1 to 1.13.0.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.12.1...v1.13.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

spolifroni-amd committed:
Docs - Update LICENSE.txt to 2025 (#259)

Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Merge branch 'develop' into crop_resize_support
